### ISSUE TITLE
Various errors with their fixes

### ISSUE THREAD

**Issue by Odyseus** [BigE/desk-changer#19]
Hello, @BigE.

While porting this extension into a Cinnamon applet, I found several errors in the daemon's code. Although these errors are not significant to you nor to your extension, I thought of reporting them since these errors also happen in any Ubuntu 16.04 based distribution using Gnome Shell.

### GLib.log_default_handler error

Using the call to this function like follows...

```python
GLib.log_default_handler(None, level, message)
```

...it will complain that the first argument shouldn't be None (even when in the documentation says that None is allowed ¬¬) and that the 4 arguments are required.

Using the following code will make it work on Ubuntu 16.04 based systems:

```python
GLib.log_default_handler('Application', level, message, 0)
```

### _emit_signal error

Using the following code...

```python
retval = GLib.VariantBuilder(variant_type_from_string("r"))
```

...will fail saying that arguments aren't supported. This will cause all signals emitted by the daemon to fail and the wallpaper preview on the menu will never update (among other problems).

Using the following code will make all work as it should:

```python
retval = GLib.VariantBuilder.new(variant_type_from_string("r"))
```

### Memory leak

Every time the daemon starts, the memory consumed by the **gnome-shell** process almost instantly doubles. And in every update of the wallpaper preview, the **gnome-shell** memory consumption goes up about 2 to 5 megabytes. I did several tests on Ubuntu Gnome 16.04 and in Archlinux with Gnome Shell and in both systems running the daemon with an interval of 10 seconds the memory consumed by the **gnome-shell** process went up to about 700Mb-1Gb in a lapse of just 2 hours.

Long story short, the daemon itself isn't the culprit. I tracked down the leak to the use of a **Clutter.Texture()**. I switched to a **Clutter.Image()** and the memory leak just disappeared.

I managed to fix this on my port for Cinnamon, but I couldn't make the **Clutter.Image()** work on Gnome Shell. To be honest, I didn't try too hard either. I simply cannot work on Gnome Shell.

**Comment by BigE** [comment]
Thanks so much for all the excellent feedback. I'm going to be working on things this weekend and will be working through some of these. This is my first attempt into application/extension development, so I'm learning as I go here :smiley: You might check out the new feature branch [feature/17-lockscreen-support](https://github.com/BigE/desk-changer/tree/feature/17-lockscreen-support) as I'll be working from here and some of the changes are already made. This will also be the next release, which is a significant rewrite of the daemon.

### Memory Leak
I've had some inkling lately that there is a huge memory leak somewhere in the extension itself. That helps immensely in tracking it down. I've always had an idea it was in the Texture, but how exactly did you determine this exactly? I've also attempted to make **Clutter.Image()** work in its place, but I have had no luck either. I'm starting to look into **St.TextureCache** but not having much luck there either. I can't imagine that **Clutter.Texture** would have such a huge memory leak in it either, perhaps there is something I'm not destroying properly?

**Comment by Odyseus** [comment]
Hello, @BigE.

Although I have several applets/extensions for Cinnamon published, I'm not a professional developer. This is just a hobby for me. I'm just very glad that I could be of some help to you.

## About the memory leak

I first realize of the memory leak immediately after I fixed the **_emit_signal** function on the daemon. I realized that it had to do something with the preview because I observed how the memory went up every time the preview was updated, so I commented the part of the code that creates and attaches the preview to the menu, and then the leak was gone.

I also looked into a better way to handle the **Clutter.Texture()** (if it could be *better destroyed* and such). But reading at the [Clutter documentation](https://lazka.github.io/pgi-docs/#Clutter-1.0/classes/Texture.html#Clutter.Texture) I found out that **Clutter.Texture()** is deprecated and it recommends the use of **Clutter.Image()** instead.

It has cost me a lot to make the **Clutter.Image()** work (like I said, I'm a newbie myself). But searching and looking at code in GitHub on how **Clutter.Image()** works helped me a lot. I recently uploaded my applet based in your extension into my repository. You can see how I made the **Clutter.Image()** work in there. Basically, the [**set_wallpaper** function](https://github.com/Odyseus/CinnamonTools/blob/master/applets/0dyseus%40WallpaperChangerApplet/files/0dyseus%40WallpaperChangerApplet/__init__.js#L572) asynchronously creates a stream of the image file ([**_createImageTexture** function](https://github.com/Odyseus/CinnamonTools/blob/master/applets/0dyseus%40WallpaperChangerApplet/files/0dyseus%40WallpaperChangerApplet/__init__.js#L513)) and then that stream is used to asynchronously create an scaled Pixbuf ([**_textureFromStream** function](https://github.com/Odyseus/CinnamonTools/blob/master/applets/0dyseus%40WallpaperChangerApplet/files/0dyseus%40WallpaperChangerApplet/__init__.js#L526)) (Yes, I got dizzy too. LOL). The three functions that I mentioned are close together, so its kind of easy to follow the code.

**Comment by BigE** [comment]
Okay, after a good weekend, I've made quite a few changes... but I'm not ready to close this yet. I did some testing on a debian vm and had some pretty bad results. I'll need to do a lot more testing to make sure this is ready before I actually make a release. I've tagged version 22 already, but I will more than likely make changes and release 23 instead.

I fixed the memory leak and switched *Clutter.Texture()* to *Clutter.Image()* ... what I kept missing before was setting the size of the actor. Once I passed that in the constructor, everything showed up and worked great. I didn't create mine the async route like you did... but I probably should. Thank you for pointing this out and helping out with it.

I'll update further once I do more investigation, probably later in the week.

**Comment by BigE** [comment]
Okay, after doing a bit more research, I'm ready to call this one understood and closed. In previous versions of GLib (<2.50) the g_log_default_handler() method did not accept NULL arguments. **HOWEVER** it appears the documentation changed to show them accepting NULL prior to the code fully changing. The annotations were added in https://github.com/GNOME/glib/commit/25a7c817d3cbef8b6508b44bca6d0b01ca77916b and the actual NULL check was added in https://github.com/GNOME/glib/commit/f414cddc1f765d06bbac9ddde0930f4567ccc70b As you can see by the tags in these changes, the annotations were added in 2.47.2 and the actual check for null wasn't added until 2.49.5. This means there is a grey area in 2.48 where the documentation says it supports NULL (or None in python) and it will cause a crash. I've updated the daemon to reflect this in e6217cd and set the first argument to "" because of how the documentation is now.

> log_domain
the log domain of the message, or NULL for the default "" application domain.

This seems to run fine on a Ubuntu 16.04.2 VM at least, which should (by default) have a version of 2.48 installed. I'm still doing more testing on Debian Stable for earlier versions of glib and such. Thanks for reporting these errors!
